### PR TITLE
Fix error in return value evaluation in ECDH_compute_key

### PR DESCRIPTION
**Pull request by schregger**
Function compute_key, called in ECDH_compute_key returns "-1" in error cases. But ECDH_compute_key checks for not "0" only.

### CODE REVIEW

**Comment by kroeckx** [comment]
Don't the other error cases get changed to return -1 too?

File: crypto/ec/ec_kmeth.c [hunk 1]
```diff
@@ -139,8 +140,10 @@ int ECDH_compute_key(void *out, size_t outlen, const EC_POINT *pub_key,
         ECerr(EC_F_ECDH_COMPUTE_KEY, EC_R_INVALID_OUTPUT_LENGTH);
         return 0;
     }
-    if (!eckey->meth->compute_key(&sec, &seclen, pub_key, eckey))
-        return 0;
+    compute_key_return = 
+        eckey->meth->compute_key(&sec, &seclen, pub_key, eckey);
+    if (compute_key_return <= 0)
+        return compute_key_return;
```
Comment (kroeckx) [inline crypto/ec/ec_kmeth.c]: I suggest you change the check from `!` to `< 0` and then just return -1, no need for the variable.

Comment (schregger) [reply crypto/ec/ec_kmeth.c]: Good point. I will change the return of the other error cases to "-1" too.

**Comment by mattcaswell** [comment]
I am confused by your assertion that:

> Function compute_key, called in ECDH_compute_key returns "-1" in error cases.

As far as I can see for all OpenSSL internal EC_KEY_METHOD implementations - the compute_key function ultimately calls `ossl_ecdh_compute_key`. That either returns 0 in error, or the return value from calling the `ecdh_compute_key` function from the EC_METHOD. Again for all internal implementations (that I can see) that ultimately calls `ecdh_simple_compute_key`

The definition of that function is here:
https://github.com/openssl/openssl/blob/master/crypto/ec/ecdh_ossl.c#L38

In all cases that functions seems to return 0 on failure. So changing things to -1 as you do in this PR seems to be the wrong thing to do. In what circumstances do you see -1 returned?

**Comment by sigiesec** [comment]
@mattcaswell At least in 1.0.2, -1 was returned at least in some error cases. For ossl_ecdh_compute_key, this was changed in https://github.com/openssl/openssl/commit/65ea288d47bd3ec8c787108d7e56d8f5ff43281f

Was this intentionally changed with 1.1.0? Is this documented somewhere?

**Comment by mattcaswell** [comment]
In 1.0.2 ECDH_compute_key sometimes returns 0 for error:

https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/ecdh/ech_key.c#L79

And sometimes it returns -1 on error:

https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/ecdh/ech_ossl.c#L125

So any caller will need to check both possible error returns (typically by checking <= 0). In 1.1.0 we have clearly attempted to standardise on one single error value 0. This is still consistent with the 1.0.2 way of doing things (anything that correctly checked for errors in 1.0.2 will still work in 1.1.0). However we have often seen code incorrectly check for errors with similar functions like this:

````
if (!ECDH_compute_key(...))
    handler error;
```

This code would not work correctly in 1.0.2 (for some errors), but it now will. 

Unfortunately ECDH_compute_key() is currently undocumented. PRs welcome to document it.

**Comment by kroeckx** [comment]
Note that DH_compute_key() documents -1 as the value for error, so
I was expecting ECDH_compute_key() to also return -1 for error.

**Comment by mattcaswell** [comment]
> Note that DH_compute_key() documents -1 as the value for error, so
I was expecting ECDH_compute_key() to also return -1 for error.

Yes, I noticed that too. But nonetheless ECDH_compute_key() from 1.1.0 onwards only retursn 0 on error. That's an unfortunate discrepancy (but there is also a discrepancy in 1.0.2 because ECDH_compute_key() can return 0 or -1 on error, and not just -1 like DH_compute_key()).

**Comment by schregger** [comment]
So, at the end we can conclude that the fix is not needed, I guess.

**Comment by mattcaswell** [comment]
I think so - closing.

**Pull request closed without merging**